src/schemas/Entity.js

```diff
@@ -49,10 +49,15 @@ export default class EntitySchema {
   }
 
   normalize(input, parent, key, visit, addEntity, visitedEntities) {
-    if (visitedEntities.some((entity) => entity === input)) {
-      return this.getId(input, parent, key);
+    const id = this.getId(input, parent, key);
+
+    if (!(id in visitedEntities)) {
+      visitedEntities[id] = [];
+    }
+    if (visitedEntities[id].some((entity) => entity === input)) {
+      return id;
     }
-    visitedEntities.push(input);
+    visitedEntities[id].push(input);
```

Comment:
Just another implementation thought, no need to act on it:

I wonder if `visitedEntities` could be a map structured by key, then id:
```js
visitedEntities = {
  [key: string]: { [id: string]: entity }
};
```

As it is now, entities of the same type are jumbled together if they have the same `id`, which kind of just makes it weird if someone were to step through and inspect it with their debugger.

Reply:
Yes, I skipped the `key` for compactness, but you’re right that would be more explicit. I can make that change if you want.

Reply:
I just did it, I hope this can be merged soon now. This causes a big performance hit for us.